p2pserver/connect_controller/connect_controller.go

```diff
@@ -0,0 +1,379 @@
+		ConnCtrlOption: option,
+		selfId:         keyid,
+		peerInfo:       peerInfo,
+		inoutbounds:    [2]*strset.Set{strset.New(), strset.New()},
+		connecting:     strset.New(),
+		peers:          make(map[common.PeerId]*connectedPeer),
+	}
+	// put domain to the end
+	sort.Slice(control.ReservedPeers, func(i, j int) bool {
+		return net.ParseIP(control.ReservedPeers[i]) != nil
+	})
+
+	return control
+}
+
+func (self *ConnectController) OwnAddress() string {
+	return self.ownAddr
+}
+
+func (self *ConnectController) getConnectId() uint64 {
+	return atomic.AddUint64(&self.nextConnectId, 1)
+}
+
+func (self *ConnectController) hasInbound(addr string) bool {
+	self.mutex.Lock()
+	defer self.mutex.Unlock()
+	return self.inoutbounds[INBOUND_INDEX].Has(addr)
+}
+
+func (self *ConnectController) OutboundsCount() uint {
+	self.mutex.Lock()
+	defer self.mutex.Unlock()
+	return uint(self.inoutbounds[OUTBOUND_INDEX].Size())
+}
+
+func (self *ConnectController) InboundsCount() uint {
+	self.mutex.Lock()
+	defer self.mutex.Unlock()
+	return uint(self.inoutbounds[INBOUND_INDEX].Size())
+}
+
+func (self *ConnectController) isBoundFull(index int) bool {
+	count := self.boundsCount(index)
+	if index == INBOUND_INDEX {
+		return count >= self.MaxConnInBound
+	}
+	return count >= self.MaxConnOutBound
+}
+
+func (self *ConnectController) boundsCount(index int) uint {
+	self.mutex.Lock()
+	defer self.mutex.Unlock()
+	return uint(self.inoutbounds[index].Size())
+}
+
+func (self *ConnectController) hasBoundAddr(addr string, index int) bool {
+	self.mutex.Lock()
+	defer self.mutex.Unlock()
+	return self.inoutbounds[index].Has(addr)
+}
+
+func (self *ConnectController) tryAddConnecting(addr string) bool {
+	self.mutex.Lock()
+	defer self.mutex.Unlock()
+
+	if self.connecting.Has(addr) {
+		return false
+	}
+	self.connecting.Add(addr)
+
+	return true
+}
+
+func (self *ConnectController) removeConnecting(addr string) {
+	self.mutex.Lock()
+	defer self.mutex.Unlock()
+
+	self.connecting.Remove(addr)
+}
+
+func (self *ConnectController) reserveEnabled() bool {
+	return len(self.ReservedPeers) > 0
+}
+
+// remoteAddr format 192.168.1.1:61234
+func (self *ConnectController) inReserveList(remoteIPPort string) bool {
+	// 192.168.1.1 in reserve list, 192.168.1.111:61234 and 192.168.1.11:61234 can connect in if we are using prefix matching
+	// so get this IP to do fully match
+	remoteAddr, _, err := net.SplitHostPort(remoteIPPort)
+	if err != nil {
+		return false
+	}
+	// we don't load domain in start because we consider domain's A/AAAA record may change sometimes
+	for _, curIPOrName := range self.ReservedPeers {
+		curIPs, err := net.LookupHost(curIPOrName)
+		if err != nil {
+			continue
+		}
+		for _, digIP := range curIPs {
+			if digIP == remoteAddr {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (self *ConnectController) checkReservedPeers(remoteAddr string) error {
+	if !self.reserveEnabled() || self.inReserveList(remoteAddr) {
```

Comment:
if self.reserveEnabled() == 0, it will return nil. but actually it need return error.

Reply:
if !self.reserveEnabled() || !self.inReserveList(remoteAddr) return err

Reply:
if reserve is not enabled, nothing need check, so return nil.

Reply:
so this function use to check what? i see you return err info is "the remote add not int reserved list",if reserve is not enabled, the remote addr is not in reserved list, so it also need return err.